ESPS/general/src/copysps/copysps.1
.\" Copyright (c) 1987-1990 Entropic Speech, Inc.
.\" Copyright (c) 1990-1996 Entropic Research Lab, Inc. All rights reserved.
.\" @(#)copysps.1	3.21 4/2/97 ESI/ERL
.ds ]W (c) 1996 Entropic Research Laboratory, Inc.
.TH COPYSPS 1\-ESPS 4/2/97
.SH NAME
copysps \- copies selected records of an ESPS file to a new ESPS file
.SH SYNOPSIS
.B copysps
[
.BI \-x " debug_level"
] [
.BI \-f
] [
.BI \-r " range"
] [
.BI \-g " gen_range"
] [
.BI \-s " time_range"
] [
.B \-z
] [
.I " infile"
]
.I " outfile"
.SH DESCRIPTION
.PP
.I Copysps
copies selected records of an ESPS file to another ESPS file.
If 
.I outfile
exists, then the selected records of
.I infile
are appended to
.I outfile
provided that the input file and the output file are compatible (see
below for details).
If 
.I outfile
doesn't exist, it is created.
Appending data to 
.I infile
is not allowed.  
If \fIoutfile\fP is "\-", then stdout is used as output.  If a single
file name is given on the command line, \fIcopysps\fP uses that name
as the output file and retrieves the input file from ESPS Common (see
EXAMPLES).
Note that reading stdin is not supported.
.PP
For FEA_SD files, additional capabilites are offered by 
\fIcopysd\fP (1\-\s-1ESPS\s+1), which permits scaling the data 
and changing the numerical data type.  
.PP
If \fIoutfile\fP already exists, then \fIcopysps\fP checks the input
file header and the output file header for compatibility.  Two ESPS
files are considered compatible if they satisfy the following
conditions:
.IP "For ESPS FEA files:"
Feature header items fea_type, field count, names, sizes, ranks,
types, and enums must be the same in both files.  
If a field is a
derived field, then the names pointed to by srcfields must also agree
in both files.  Since \fIcopysps\fP does not require compatibility of
generic header items (indeed, none are copied from the input to the
output file if the output file exists already), you should be careful
when copying records of FEA_SD, FEA_ANA, FEA_SPEC, FEA_STAT, and FEA_VQ files.
Refer to section 5 man pages for more information about 
these special feature files. 
.IP "For ESPS FEA_SD files:"
The sampling rate of both files must be the same.  
.IP "For ESPS SPEC files:"
Spectral header items order_vcd, order_unvcd, spec_an_meth, dcrem,
voicing, freq_format, spec_type, and contin must all be the same for
both files.  If win_type, sf, spec_an_meth, or dcrem are not the same
in both files, then a warning message is printed on stderr and in the
comment field of the output file.  (Note: SPEC files are obsolete 
and have been replaced by FEA_SPEC files.)  
.IP "For ESPS FILT files:"
Filter header items max_num, max_den, func_spec, nbands, npoints,
nbits, type, bandedges, points, gains, and wts must all be the same
in both files.
.IP "ESPS SCBK files not supported, yet."
.PP
When updating an existing file without the \fB-f\fR option, a
temporary file is generated in the directory specified by the
environment variable ESPS_TEMP_PATH (default /usr/tmp).  The operation
will be faster if the destination file is on the same file system as
ESPS_TEMP_PATH.  This is not a significant difference unless the
destination file is very large.
.SH OPTIONS
The following options are supported:
.TP
.BI \-f
This option causes a fast copy to be done.  This is done by appending
the selected records onto the end of the destination file.
Use of this option causes embedded headers not to be included in the
output file.  Also, the command line is not saved as a comment in the
output header (the output header size cannot change).  However, note
that \fIcomment\fP(1\-ESPS) can always be used to add comments to an
existing ESPS file.
This option should not be used to produce archival files where the
entire processing history is needed.
.br
.sp
If the header version (\fIcommon.hdvers\fR) is not the same for the
input and output file, then a fast copy is not done.  This is because
the size or the exact format of the header might be different and the
write may fail.   The program checks the versions, and simply turns
the \fB\-f\fR flag off if the versions do not match.
.sp
Note that two segment labeled files are not compatible unless they both
refer to the same set of sampled data files.  Also, if you use copysps
to copy tagged records to a tagged file, it will become difficult to
determine which file the output tags refer to in cases where the input
file has tags refering to a different file than the tags in the output
file.  To deal with such problems for now, use select instead of
copysps.
.TP
.BI \-r " start:end"
.TP
.BI \-r " start:+incr
Determines the range of data records to process.  In the first form, a
pair of unsigned integers gives the first and last points of the range.  
If 
.I start
is omitted, 1 is used.  If 
.I end
is omitted, the last point in the file is used.  The second form is
equivalent to the first with 
.I "end = start + incr".
.TP
.BI \-g " range"
Select a "generic" range of records to be processed.  The default is
all the records in \fIinfile\fP. (See
.I grange_switch(3\-ESPS) 
for full details of generic range specification.)
.TP
.BI \-s " start_time:end_time"
.TP
.BI \- " start_time:+time_incr
Determines the range of records  to be copied from \fIinfile\fP by 
in units of time (seconds).  In the first form, a pair of
signed real numbers gives the start and end time of the range. If 
.I start_time
is omitted, the first record is used. 
.I end_time
is omitted, the last record in the file is used.  The second form is
equivalent to the first with 
.I "end_time = start_time + time_incr".
.TP
.BI \-x " debug_level"
Only debug level 1 is defined in this version;
this causes several messages to be printed.
The default level is zero, which causes no debug output.
.TP
.B \-z
Suppresses messages that inform the user when any of the input
filename, the starting record, or number of records is taken from 
ESPS Common (see ESPS COMMON).  
.SH ESPS PARAMETERS
The ESPS parameter file is not read.
.SH ESPS COMMON
.PP
ESPS Common processing may be disabled by setting the environment variable
USE_ESPS_COMMON to "off".  The default ESPS Common file is .espscom 
in the user's home directory.  This may be overidden by setting
the environment variable ESPSCOM to the desired path.  User feedback of
Common processing is determined by the environment variable ESPS_VERBOSE,
with 0 causing no feedback and increasing levels causing increasingly
detailed feedback.  If ESPS_VERBOSE is not defined, a default value of 3 is
assumed.
.PP
The following items are read from the ESPS Common File provided that
only one input file or no input file is given on the command line and
provided that standard input isn't used.  
.IP
.I "filename \- string"
.IP
This is the name of the input file.  If no input file is 
specified on the command line, \fIfilename\fP is taken to be the
input file.  If an input file is specified on the command line, 
that input file name must match \fIfilename\fP or the other 
items (below) are not read from Common. 
.IP
.I "start \- integer"
.IP
This is the starting point in the input file to begin copying.  It 
is not read if the \fB\-r\fP option is used.  
.sp
.I "nan \- integer"
.IP
This is the number of points to copy from the input file.  It is not
read if the \fB\-r\fP option is used.  A value of zero means the last
point in the file.
.sp
.PP
Again, the values of \fIstart\fR and \fInan\fR are only used if the
input file on the command line is the same as \fIfilename\fP in the
common file, or if no input file was given on the command line.  If
\fIstart\fR and/or \fInan\fR are not given in the common file, or if
the common file can't be opened for reading, then \fIstart\fR
defaults to the beginning of the file and \fInan\fR defaults to the
number of points in the file.
.PP
The following items are written into the ESPS Common file:
.IP
.I "start \- integer"
.IP
The starting point from the input file.
.sp
.I "nan \- integer"
.IP
The number of points in the selected range.
.sp
.I "prog \- string"
.IP
This is the name of the program (\fIcopysps\fP in this case).
.sp
.I "filename \- string"
.IP
The name of the input file.
.PP
These items are not written to ESPS COMMON if the output file is <stdout>.
.SH ESPS HEADER
If the output file is new (i.e. not appending to an existing file)
then the source ESPS header structure is copied into the new file
header (including generic header items).  For exiting output files,
the input ESPS file name is added to the list of sources.  The command
line and compatibility warnings, if any, are added in the comment
field of the header.  
.PP
Also, if the output file is new and the generic header item
\fIrecord_freq\fP exists in the input file, the generic header item
\fIstart_time\fP is written in the output file.  The value written is
computed by taking the \fIstart_time\fP value from the header of the
input file (or zero, if such a header item doesn't exist) and adding
to it the offset time (from the beginning of the input file) of the
first record processed.  If \fIrecord_freq\fP doesn't exist in the
input file, \fIstart_time\fP is not written in the output file.  If it
exists in the input file header, the generic header item
\fIrecord_freq\fP is copied to the output file header.  This item
gives the number of records per second of original data analyzed.
.PP
If the output file already exists and it is is being appended to,
\fIstart_time\fP is not modified and the \fIrecord_freq\fP
values in the two files are checked for agreement. If they agree,
no change is made to it; If they don't agree, the \fIrecord_freq\fP
value is changed to zero.
.PP
In update mode, the input and the output files must both be in EDR (Entropic's
external data representation) or the the machine's native format.
.SH EXAMPLES
.PP
To copy records 1 to 5, 9, and 12 to 15 from \fIfile1.sps\fP to
\fIfile2.sps\fP (assuming \fIfile2.sps\fP does not already exist),
type the following:
.sp
.nf
    % copysps -r1:5,9,12:15 file1.sps file2.sps
.fi
.sp
Note that \fIfile1.sps\fP and \fIfile2.sps\fP can be any ESPS file.
Only ESPS FEA, FILT, SPEC, or SCBK files can be updated (i.e. have data
appended to them).  Continuing our example above, if \fIfile2.sps\fP
is an ESPS FEA file, then we can append records from \fIfile3.sps\fP
by typing:
.sp
.nf
    % copysps file3.sps file2.sps
.fi
.sp
If only \fIfile2.sps\fP is given on the command line, as in the following
example:
.sp
.nf
    % copysps file2.sps
.fi
.sp
then \fIcopysps\fP gets the input file name from ESPS Common.
.SH DIAGNOSTICS
.PP
A fatal error occurs if the input file does not exist, if it is not an ESPS
file, or if a requested range includes records that do not exist.
.PP
.nf
copysps: no output file specified.
Usage: copysps [-x debug] [\-f] [\-r gen_range] [-z] [infile] outfile
copysps: infile and outfile cannot be the same.
copysps: input file name \fIinfile\fP taken from ESPS Common.
copysps: fea_compat: feature fields in \fIinfile\fP and \fIoutfile\fP incompatible.
copysps: cannot handle ESPS SCBK file type.
copysps: cannot update ESPS file type code: \fItype_code\fP
copysps: could not open \fIoutfile\fP for appending.
copysps: could not open \fItemp_file\fP
copysps: could not open \fIoutfile\fP for writing.
copysps: only \fInum_rec\fP records in \fIinfile\fP
copysps: record sizes in \fIinfile\fP and \fIoutfile\fP are different.
copysps: calloc: could not allocate memory for dbuf.
copysps: read error on \fIinfile\fP
copysps: write error on \fItemp_file\fP
copysps: seek error on \fIinfile\fP
copysps: write error on \fIoutfile\fP
copysps: could not rename \fItemp_file\fP to \fIoutfile\fP
.fi
If the versions of the headers do not match when the \fB\-f\fR flag is
used, a message is printed and the \fB\-f\fR flag is turned off.
.SH SEE ALSO
.PP
.nf
\fIespsenv\fR (1\-\s-1ESPS\s+1), \fImergefea\fP (1\-\s-1ESPS\s+1), \fIaddfea\fP (1\-\s-1ESPS\s+1),
\fIcopysd\fR(1\-ESPS), \fIcomment\fR(1\-ESPS), \fIselect\fR(1\-ESPS),
\fIgrange_switch\fR(3\-ESPSu), fea(5\-ESPS), fea_ana(5\-ESPS),
fea_stat(5\-ESPS), fea_vq(5\-ESPS), filt(5\-ESPS), 
scbk(5\-ESPS), spec(5\-ESPS)
.fi
.SH BUGS
.PP
Owing to the way that \fIgrange_switch\fR is used, this might fail on very
large files.
.PP
When a new output file is created, it will always be the same data
format (either NATIVE or EDR) as the input.  This is considered a bug
and will be changed in a future version.  Note that you can use \fIselect\fR to copy
the file and convert from NATIVE to EDR.
.PP
\fICopysps\fP does not work correctly if the input and output files
are not both in field order or both in type order.  For example, if
the input file is in type order (the default) and the environment
variable FIELD_ORDER is set to "on" before running \fIcopysps\fP, the
output file will be garbage.  
.PP
Compressed Sphere files are not supported.  Files in Esignal format are
not supported.
.SH FUTURE CHANGES
.PP
Handle tagged and segment labeled FEA files properly.  
.PP
Revise to handle MIIO properly.  
.SH AUTHOR
.PP
Ajaipal S. Virdy, Entropic Speech, Inc.


ESPS/general/src/copysps/copysps.c
/*
 * This material contains unpublished, proprietary software of Entropic
 * Research Laboratory, Inc. Any reproduction, distribution, or publication
 * of this work must be authorized in writing by Entropic Research
 * Laboratory, Inc., and must bear the notice:
 * 
 * "Copyright (c) 1987-1990  Entropic Speech, Inc. "Copyright (c) 1990-1996
 * Entropic Research Laboratory, Inc. All rights reserved"
 * 
 * The copyright notice above does not evidence any actual or intended
 * publication of this source code.
 * 
 * Written by:  Ajaipal S. Virdy Checked by: Revised by: Alan Parker
 * 
 * Copy records of an ESPS file to a new ESPS file or append to an existing ESPS
 * file.
 * 
 */

#ifdef SCCS
static char    *sccs_id = "@(#)copysps.c	3.24	6/3/98	ESI/ERL";
#define VERSION "3.24"
#define DATE "6/3/98"
#endif

#ifndef VERSION
#define VERSION "debug"
#endif
#ifndef DATE
#define DATE "none"
#endif

#include <stdio.h>
#include <errno.h>

#include <esps/esps.h>
#include <esps/fea.h>
#include <esps/feasd.h>
#include <esps/spec.h>
#include <esps/filt.h>
#include <esps/scbk.h>
#include <esps/unix.h>
#include <esps/esignal_fea.h>


#define SYNTAX USAGE ("copysps [-x debug_level] [-f] [-r|-p gen_range] [-s time_range ] [-z] [infile] outfile")
#define ERROR_EXIT(text) {(void) fprintf(stderr, "copysps: %s - exiting\n", text); exit(1);}

int             debug_level = 0;

extern void     trange_switch();
char           *e_temp_name();
char           *get_sphere_hdr();

main(argc, argv)
	int             argc;
	char          **argv;
{				/* begin main */

	int             c, getopt();
	extern int      optind;
	extern int      errno;
	extern char    *optarg;

	char           *get_cmd_line();

	char           *f1name, *f2name, *tmpname;
	FILE           *f1strm, *f2strm, *tstrm;

	struct header  *f1hd, *f2hd;

	char           *dbuf;	/* array to hold data read in */

	int             debug = 0;
	int             rflag = 0;
	int             gflag = 0;
	long           *rec_range = NULL;
	long            tot_rec;
	int             sflag = 0;
	char           *srange = NULL;
	char           *rrange = NULL;
	long            input_ndrec = 0;

	long            i, s_rec = 1, e_rec;

	int             fast = 0;
	int             nohead = 0;
	int             update = 0;	/* set if we have to update output
					 * file */
	int             nofile = 0;	/* set if no input file on command
					 * line */
	int             zflag = 0;

	long            cur_pos, new_pos;	/* variables for positioning
						 * seek() */


	char           *template = "copyspsXXXXXX";
	struct fea_data *f1rec, *f2rec;

	/* ESPS functions referenced */

	long           *grange_switch();
	int             size_rec();
	int             recsize1, recsize2;
	char           *eopen();
	char           *savestring();
	int             j;
	/* Local functions referenced */

	void            check_sd();
	void            check_spec();
	void            check_filt();
	void            add_sd_comment();
	void            add_spec_comment();
	void            add_filt_comment();
	long            common_nan;
	while ((c = getopt(argc, argv, "x:fr:p:s:hzg:")) != EOF)
		switch (c) {
		case 'x':
			debug = atoi(optarg);
			break;
		case 'p':
		case 'g':
			rec_range = grange_switch(optarg, &tot_rec);
			gflag = 1;
			break;
		case 's':
			srange = optarg;
			sflag = 1;
			break;
		case 'r':
			rrange = optarg;
			rflag = 1;
			break;
		case 'f':
			fast++;
			break;
		case 'h':
			nohead++;
			break;
		case 'z':
			zflag++;
			break;
		default:
			SYNTAX;
		}

	/* Determine if output file is specified */
	if (argc - optind < 1) {
		Fprintf(stderr, "copysps: no output file specified.\n");
		SYNTAX;
	}
	/*
	 * If f2name is "-", then use stdout for f2strm. If f2name exists,
	 * then update the file, otherwise create it.
	 */

	f2name = argv[argc - 1];
	if (strcmp(f2name, "-") != 0)
		if ((f2strm = fopen(f2name, "r")) == NULL)
			update = 0;	/* outfile doesn't exist */
		else
			update = 1;	/* outfile exists */

	if (argc - optind < 2) {/* only output file name specified */
		/* user must supply two file names (or "-") */
		/* SYNTAX; */
		nofile = 1;
		f1name = NULL;
	} else {		/* open input file */
		f1name = eopen("copysps", argv[optind], "r", NONE, NONE,
			       &f1hd, &f1strm);
		if (strcmp(f1name, "<stdin>") == 0) {
			ERROR_EXIT("Standard input not supported");
		}
	}


	/*
	 * If command line has no input file name, then it might be in ESPS
	 * common; we need to read_params anyway (provided stdin not used),
	 * so we do it here (checking common only); However, we don't check
	 * common if more than one input file is given since this is error
	 * prone.
	 */

	if (nofile || ((f1strm != stdin) && (argc - optind <= 2))) {
		if (debug)
			Fprintf(stderr, "copysps: checking ESPS common\n");
		(void) read_params((char *) NULL, SC_CHECK_FILE, f1name);

	}
	if (nofile) {
		if (symtype("filename") == ST_UNDEF) {
			ERROR_EXIT("no input file name on command line or in common");
		} else {
			f1name = getsym_s("filename");

			if (debug)
				Fprintf(stderr,
					"copysps: input file name from common is %s\n",
					f1name);

			if (strcmp(f2name, f1name) == 0)
				ERROR_EXIT("input name from common same as output file");

			(void) eopen("copysps", f1name, "r",
				     NONE, NONE, &f1hd, &f1strm);
			if (!f1hd->common.edr && edr_default(f1hd->common.machine_code))
				f1hd->common.edr = YES;
			if (!zflag)
				Fprintf(stderr,
					"copysps: input file name %s taken from ESPS Common.\n",
					f1name);
		}
	}
	s_rec = 1;
	e_rec = (f1hd->common.ndrec != -1 ? f1hd->common.ndrec : LONG_MAX);

	if (rflag)
		range_switch(rrange, &s_rec, &e_rec, 1);

	if (sflag) {
		s_rec = 1;
		e_rec = f1hd->common.ndrec;
		trange_switch(srange, f1hd, &s_rec, &e_rec);
		if (s_rec < 0) {
			ERROR_EXIT("requested start time is less than start_time of file");
		}
	}
	if (!gflag && !sflag && !rflag) {
		s_rec = 1;
		e_rec = f1hd->common.ndrec;
		if (symtype("start") != ST_UNDEF) {
			s_rec = getsym_i("start");
			if (!zflag)
				Fprintf(stderr,
					"copysd: starting record = %ld taken from ESPS Common\n",
					s_rec);
		}
		if (symtype("nan") != ST_UNDEF) {
			common_nan = getsym_i("nan");
			if (!zflag)
				Fprintf(stderr,
					"copysd: number records = %ld taken from ESPS Common\n",
					common_nan);
			if (common_nan != 0)
				e_rec = s_rec + common_nan - 1;
		}
	}
	if (strcmp(f1name, f2name) == 0) {
		ERROR_EXIT("infile and outfile cannot be the same");
	}
	if (debug)
		Fprintf(stderr,
			"copysps: infile = %s, outfile = %s, update = %d, fast = %d\n",
			f1name, f2name, update, fast);

	/* Check if outfile exists and is an ESPS file */

	if (update) {		/* outfile exists */

		if (!(f2hd = read_header(f2strm)))
			NOTSPS("copysps", f2name);
		if (get_esignal_hdr(f2hd)) {
			ERROR_EXIT("cannot append to an Esignal file");
		}
		if (get_sphere_hdr(f2hd)) {
			ERROR_EXIT("cannot append to a Sphere file");
		}
		if (get_pc_wav_hdr(f2hd)) {
			ERROR_EXIT("cannot append to a PC WAV (RIFF) file");
		}
		input_ndrec = f2hd->common.ndrec;

		/* Check for compatibility */

		if (debug)
			Fprintf(stderr,
				"copysps: %s already exists; check for compatibility.\n",
				f2name);

		switch (f2hd->common.type) {
		case FT_FEA:{
				short         **trans;	/* translation table
							 * returned */
				int             com_lev;	/* compatibility level */

				if (debug)
					Fprintf(stderr,
					  "copysps: calling fea_compat.\n");

				com_lev = fea_compat(f1hd, f2hd, (char **) NULL, &trans);

				if (debug)
					Fprintf(stderr,
						"copysps: com_lev = %d, trans is %s.\n", com_lev,
						(trans == NULL) ? "NULL" : "not NULL");

				if ((com_lev != 1) || (trans != NULL)) {
					Fprintf(stderr,
						"copysps: fea_compat: feature fields in %s and %s incompatible.\n",
						f1name, f2name);
					exit(1);
				}
				/*
				 * check record_freq, if it exists and files
				 * FEA_SD
				 */
				if (f2hd->hd.fea->fea_type == FEA_SD) {
					if (get_genhd_val("record_freq", f1hd, (double) 0) !=
					    get_genhd_val("record_freq", f2hd, (double) 0)) {
						Fprintf(stderr,
							"copysps: %s and %s have different record_freq - exiting.\n", f1name, f2name);
						exit(1);
					}
				}
			}
			break;
		default:
			Fprintf(stderr,
			 "copysps: cannot update ESPS file type code: %d\n",
				f1hd->common.type);
			exit(1);
			break;
		}		/* end switch (f2h->common.type) */

		if (debug)
			Fprintf(stderr,
				"copysps: input files compatible; now check for fast mode.\n");

		if (fast) {
			/*
			 * fast flag is set, append selected records to the
			 * output file.
			 */
			if (debug)
				Fprintf(stderr,
					"copysps: fast mode set: append to %s.\n", f2name);
			(void) fclose(f2strm);
			if ((f2strm = fopen(f2name, "a+")) == NULL) {
				/*
				 * ugly hack to fix r+ on SYS 5.2
				 * tstrm->_flag |= _IOWRT;
				 */
				Fprintf(stderr,
					"copysps: could not open %s for appending.\n", f2name);
				exit(1);
			}
		} else {
			/*
			 * We have to update outfile, but not in fast mode!
			 * This means that we have to retain all header
			 * information and include as a subheader the header
			 * of infile. Therefore, create a temporary file to
			 * store results.
			 */
			tmpname = e_temp_name(template);
			if ((tstrm = fopen(tmpname, "w")) == NULL) {
				Fprintf(stderr,
				   "copysps: could not open %s\n", tmpname);
				exit(1);
			}
			if (debug)
				Fprintf(stderr,
					"copysps: fast mode not set: creating tmp file %s.\n",
					tmpname);
		}		/* end if (fast) */

		/*
		 * check record_freq values
		 */
		{
			double          record_freq = 0;
			if (get_genhd_val("record_freq", f1hd, (double) 0) !=
			    get_genhd_val("record_freq", f2hd, (double) 0))
				(void) add_genhd_d("record_freq", &record_freq,
						   (int) 1, f2hd);
		}

	}
	/* end outfile exists */
	else {			/* outfile doesn't exist */

		if (debug)
			Fprintf(stderr, "copysps: %s does not exit, create it.\n",
				f2name);

		if (strcmp(f2name, "-") == 0) {
			f2strm = stdout;
			f2name = savestring("<stdout>");
		} else if ((f2strm = fopen(f2name, "w")) == NULL) {
			Fprintf(stderr,
			"copysps: could not open %s for writing.\n", f2name);
			exit(1);
		}
		fast = 0;	/* reset fast flag if on */
		f2hd = copy_header(f1hd);
		f2hd->common.edr = f1hd->common.edr;

		if (debug)
			Fprintf(stderr,
			     "copysps: copied infile header to outfile.\n");

		/*
		 * update start time and record freq
		 */
		if (rec_range == NULL)
			update_waves_gen(f1hd, f2hd, (double) s_rec, 1.0);
		else
			update_waves_gen(f1hd, f2hd, (double) rec_range[0], 1.0);


	}			/* end if (update) */

	if (!fast) {
		if ((debug) && !nohead)
			Fprintf(stderr,
			"copysps: adding source file header to outfile.\n");
		if (nohead)
			(void) add_source_file(f2hd, f1name, (struct header *) NULL);
		else
			(void) add_source_file(f2hd, f1name, f1hd);
	}
	if (debug)
		Fprintf(stderr,
			"copysps: s_rec: %d, e_rec: %d\n", s_rec, e_rec);

	if (gflag)
		if (rec_range[tot_rec - 1] > e_rec) {
			Fprintf(stderr, "copysps: only %d records in %s\n",
				f1hd->common.ndrec, f1name);
			exit(1);
		}
	recsize1 = size_rec(f1hd);
	if (recsize1 == -1) {
		ERROR_EXIT("variable record size not supported")
	}
	(void) skiprec(f1strm, s_rec - 1, recsize1);

	if (!gflag)
		tot_rec = e_rec - s_rec + 1;

	if (debug)
		Fprintf(stderr,
			"copysps: recsize1 = %ld, tot_rec = %ld\n", recsize1, tot_rec);

	if (!fast) {		/* retain all new header information in
				 * outfile */
		(void) strcpy(f2hd->common.prog, "copysps");
		(void) strcpy(f2hd->common.vers, VERSION);
		(void) strcpy(f2hd->common.progdate, DATE);
		if (!nohead) {
			(void) add_comment(f2hd, get_cmd_line(argc, argv));
		}
		/*
		 * If we have to update an existing file, then we should add
		 * any warnings in the comment field if any inconsistencies
		 * occur.
		 */

		if (update)
			switch (f2hd->common.type) {
			case FT_FEA:	/* do nothing */
				break;
			default:
				Fprintf(stderr,
					"copysps: cannot update ESPS file type code: %d\n",
					f1hd->common.type);
				exit(1);
				break;
			}	/* end switch (f2h->common.type) */

		/*
		 * If in update mode, then we must write header information
		 * out into a temporary file (pointed to by tstrm), otherwise
		 * just write it into our new outfile (pointed to by f2strm).
		 */

		if (debug)
			Fprintf(stderr,
				"copysps: writing header to %s stream.\n",
				(update == 0) ? f2name : tmpname);

		if (update)
			(void) write_header(f2hd, tstrm);
		else
			(void) write_header(f2hd, f2strm);
	}
	if (debug)
		Fprintf(stderr, "copysps: Process records\n");

	/*
	 * If we have to update outfile and we're not in fast mode, then we
	 * have to do the following: Copy all records from outfile into a
	 * temporary file.  Then append to the temporary file the the new
	 * records from infile.  Finally, we'll have to get everything from
	 * the temporary file back into outfile.
	 */

	if (update && !fast) {	/* append mode */
		long            tmp_rec;
		struct fea_data *tmpbuf;

		if (debug)
			Fprintf(stderr,"copysps: to temp ");
		recsize2 = size_rec(f2hd);
		if (recsize2 == -1) {
			ERROR_EXIT("variable record size not supported")
		}
		if (recsize2 != recsize1) {
			Fprintf(stderr,
				"copysps: record sizes in %s and %s are different.\n",
				f1name, f2name);
			exit(1);
		}
		tmp_rec = f2hd->common.ndrec;
		tmpbuf = allo_fea_rec(f2hd);
		for (i = 0; i < tmp_rec; i++) {
			if (get_fea_rec(tmpbuf, f2hd, f2strm) < 1) {
				Fprintf(stderr, "copysps: read error on %s\n", f2name);
				exit(1);
			}
			put_fea_rec(tmpbuf, f2hd, tstrm);
			if (debug && (i%1000 == 0)) fputc('#',stderr);
		}
		if(debug) fputc('\n',stderr);
		(void) free_fea_rec(tmpbuf);
	}			/* end if (update && !fast) */
	/*
	 * The pointer in the input stream is now set to the correct record
	 * position (because of read_header and skiprec). 
	 */
	cur_pos = ftell(f1strm);/* current position in f1strm (record 1) */

	/* for (i = 0; i < tot_rec; i++) */

	i = 0;
	f1rec = allo_fea_rec(f1hd);
	f2rec = allo_fea_rec(f2hd);
	if (debug) 
		Fprintf(stderr,"copysps: to temp or output ");
	while (i < tot_rec) {

		if (gflag) {
			new_pos = (rec_range[i] - 1) * recsize1 + cur_pos;

			if ((j = fseek(f1strm, new_pos, 0)) != 0) {
				break;
				/*
				 * Fprintf (stderr, "copysps: seek error on
				 * %s\n", f1name); exit (1);
				 */
			}
		}
		if (get_fea_rec(f1rec, f1hd, f1strm) < 1) {
			break;
			/*
			 * Fprintf (stderr, "copysps: read error on %s\n",
			 * f1name); exit (1);
			 */
		}
		copy_fea_rec(f1rec, f1hd, f2rec, f2hd, NULL, NULL);
		if (update && !fast) {
			put_fea_rec(f2rec, f2hd, tstrm);
		} else {
			put_fea_rec(f2rec, f2hd, f2strm);
		}
		if (debug && (i%1000 == 0)) fputc('#',stderr);
		i++;

	}			/* end while (i < tot_rec) */

	/* i is the total number of records processed */

	if (debug)
		Fprintf(stderr,
			"\ncopysps: number of records processed = %d\n", i);

	if (update && !fast) {
		if (rename(tmpname, f2name) != 0) {
			if (errno != EXDEV) {
				Fprintf(stderr,
					"copysps: could not rename %s to %s\n", tmpname, f2name);
				exit(1);
			} else {
				int i;
				if (debug) 
					Fprintf(stderr,"copysps: from temp ");
				(void) fclose(tstrm);
				if ((tstrm = fopen(tmpname, "r")) == NULL) {
					Fprintf(stderr, "copysps: cannot reopen temp file!\n");
					exit(1);
				}
				(void) unlink(tmpname);
				if ((f2strm = fopen(f2name, "w")) == NULL) {
					Fprintf(stderr, "copysps: cannot open %s\n", f2name);
					exit(1);
				}
				i=0;
				while ((c = fgetc(tstrm)) != EOF) {
					(void) fputc(c, f2strm);
					if (debug && (i++%1000 == 0)) { 
						fputc('#',stderr);
					}
				}
				if(debug) 
					Fprintf(stderr,
					   "\ncopysps: %d bytes processed\n",i);
			}
		}
		(void) fclose(tstrm);
	}
	(void) fclose(f1strm);
	(void) fclose(f2strm);

	/* put output file info in ESPS common */

	if (strcmp(f2name, "<stdout>") != 0) {
		(void) putsym_s("filename", f2name);
		(void) putsym_s("prog", argv[0]);
		(void) putsym_i("start", 1);
		(void) putsym_i("nan", (int) (tot_rec + input_ndrec));
	}
	exit(0);
	/* NOTREACHED */

}				/* end main() */
